src/rgw/rgw_rados.cc

```diff
@@ -1914,6 +1919,12 @@ int RGWRados::Bucket::List::list_objects_ordered(int64_t max_p,
   } // read attempt loop
 
 done:
+  /* XXX reject short results--n.b., count==0 */
+  if (result->size() == 0) {
+    if (restarts++ < allowed_restarts)
+      goto restart;
```

Comment:
- Do we want to restart if truncated is false?
- Also, I'm not clear why we're going to "restart:" rather than iterating again at the "attempt" for loop.
- And if we restart twice and still have 0 items, do we have the same issue?

Reply:
1. truncate:  I think so
2. I think it matters, but I'll check why I did that
3. yes, we do, that's the main thing I needed your help to think through.  I think we might need an arbitrary number of restarts--we need to ensure that marker is advancing with every read cycle, I believe